config/config.exs

```diff
@@ -0,0 +1,94 @@
+import Config
+
+## Example configuration below, set actual configuration in config/#{Mix.env()}.exs.
+#
+# config :nimble_hex,
+#   port: 4001,
+#   url: "http://localhost:4001"
+#
+# config :ex_aws,
+#   access_key_id: System.fetch_env!("NIMBLE_HEX_S3_ACCESS_KEY_ID"),
+#   secret_access_key: System.fetch_env!("NIMBLE_HEX_S3_SECRET_ACCESS_KEY"),
+#   json_codec: Jason
+#
+# config :nimble_hex,
+#   repositories: [
+#     # Sample repository with local storage
+#     #
+#     # Configure repository:
+#     #
+#     #     mix hex.repo add repo1 http://localhost:4000/repos/repo1 --public-key public_key.pem
+#     #
+#     # Configure publishing:
+#     #
+#     #     mix hex.config api_url http://localhost:4000/api/repos/repo1
+#     #
+#     repo1: [
+#       private_key: "...",
+#       public_key: "...",
+#       store: {NimbleHex.Store.Local, root: {:nimble_hex, "data"}}
+#     ],
+#     # Sample repository with S3 storage,
+#     #
+#     # Configure repository:
+#     #
+#     #     mix hex.repo add repo2 https://<bucket>.s3.<region>.amazonaws.com/repos/repo2 --public-key public_key.pem
+#     #
+#     # Remember to set bucket as public when appropriate, configure CDN in front etc.
+#     #
+#     # Configure publishing:
+#     #
+#     #     mix hex.config api_url http://localhost:4000/api/repos/repo2
+#     #
+#     repo2: [
+#       private_key: "...",
+#       public_key: "...",
+#       store:
+#         {NimbleHex.Store.S3,
+#          bucket: System.fetch_env!("NIMBLE_HEX_S3_BUCKET"),
+#          region: System.fetch_env!("NIMBLE_HEX_S3_REGION")}
+#     ],
+#     # A Mirror checks upstream repository with given frequency for changes and syncs them.
+#     #
+#     # Mirrors are read-only and must be configured with upstream URL and public key.
+#     #
+#     # The optional `:only` option configures an allowed packages list to fetch.
+#     # When using `:only` option, you need to manually make sure that all of package's
```

Comment:
I added the `:only` option mostly for testing this locally, but I can also see people using it for poor man's curation of packages.

What do you think about this option? It's perhaps kinda weird that on one hand repository says in `/names` that it has a bunch of packages, but it might effectively have just a small subset. An alternative is to have mirror maintain it's own resource files that match the actual packages. In that case, however, the mirror would have to sign the resources with it's own private key, so the mirror would no longer work out of the box with Hex.pm public key that ships with the client.

Reply:
I think having a partial mirror of only packages you use is a very good use case. Will the mirror re-sign resources, if that's the case it can build its own `names` and `versions`?

Reply:
My plan is to ship with a partial mirror like above which *does not* re-sign resources (and have caveats mentioned above), but folks could write an alternate mirror that does re-sign. WDYT?

Reply:
We have three use-cases for nimble_hex as I see it:

  1. Run it only as a mirror of hexpm
  2. Run it only as its own repository
  3. Do both of the above

For 1. we don't have to re-sign. For 2. we need to sign our own resources. How will 3. work though? Will it run two distinct repositories, one for the mirror and one for its own – or is the mirror merged into the same repository?

Reply:
Oh, I should improve documentation around this, nimble_hex supports multiple repositories out of the box, the URLs always have a repo in the path:

```
GET /repos/hexpm_mirror/names
GET /repos/my_repo/names
GET /repos/hexpm_mirror/names
POST /api/repos/my_repo/publish
```

btw, I was thinking of introducing some default repo config so that a plug can rewrite the `GET /names` request into `GET /repos/default_repo/names` but we don't need it yet.

Reply:
Perfect! Then don't re-sign and add the caveat about names and versions imo.